AB#694: Graphene: LD_PRELOAD the Marblerun Premain

Nirusu opened the pull request:
Hopefully, I didn't miss a case here where the LD_PRELOAD version has a different behavior from the separate premain executable. But so far, it looks good I guess.

For testing purposes, I already uploaded the graphene-premain.so resulting from this build to our release assets so that the Makefile already works and can download a precompiled version built from this branch. If we change some stuff here, we need to make sure that we change the release asset under the 0.3.0 release, too.

I'm not _quite_ happy with the documentation yet. I am not entirely sure if it is understandable for someone new to Marblerun what's happening here. Also, we can discuss if `auto-preload` and `manual-preload` are good names for the difference between the two or if there are better names and/or descriptions.

The -trimpath commit is mostly unrelated to this PR, but I thought I could smuggle it in here 'cause it annoyed me for quite some time now 😛

thomasten commented on samples/graphene-hello/README.md:
````diff
@@ -1,28 +1,6 @@
-make
-```
-Then get `mr_enclave` from the build output and set it as `UniqueID` in `manifest.json`.
+In essence, you have to add the `premain` process either by manually adding the Marblerun premain process as the entry point for your application (see: `manual-preload`), or by automatically injecting it as a preloaded shared library via Graphene's LD_PRELOAD feature (see: `auto-preload`).
````
We may name them `fork` (or `spawn`) and `preload`, or `fork-method` and `preload-method`. Then describe the trade-offs (additional process that causes slow start vs argv cannot be provisioned).

Nirusu commented on samples/graphene-hello/auto-preload/hello.manifest.template:
```diff
@@ -0,0 +1,43 @@
+
+# run Marblerun premain
+loader.env.LD_PRELOAD = "./premain-graphene.so"
+loader.env.LD_LIBRARY_PATH = "lib:/lib/x86_64-linux-gnu:/usr/lib:/usr/lib/x86_64-linux-gnu"
```
For the sake of this demo, properly not all of them. Only those `ldd` outputs to us.

Not sure though if hardcoding every single library makes a good demo if someone wants to use this example to load their own application in there, as it could require more shared libraries than the ones we provide.

thomasten replied:
I don't think that this is how it works in Graphene. All required libs are in the folder mounted at /lib, except premain-graphene.so, but this one has an explicit path (./...). It actually works with just "/lib" on my VM. Can you double-check?

Nirusu replied:
Does indeed work (now?). Will reduce it back to /lib.

Nirusu commented on samples/graphene-hello/manual-preload/manifest.json:
```diff
@@ -0,0 +1,19 @@
+    "Marbles": {
+        "hello": {
+            "Package": "world",
+            "Parameters": {
```
I'll go for both cases, they should not differ except when needed due to technical restrictions imo.

Nirusu commented on samples/graphene-hello/manual-preload/README.md:
```diff
@@ -0,0 +1,47 @@
+# Graphene sample (manual-preload)
+This sample shows how to run a [Graphene](https://github.com/oscarlab/graphene) application in Marblerun. In essence, you have to add the `premain` process to the Graphene manifest. `premain` will contact the Coordinator, set up the environment, and run the actual application. See the commented [hello.manifest.template](hello.manifest.template) for details.
+## Requirements
```
I think we can do this in general, though I would highlight it here in the sample as this is likely the first point an user will touch before reading the documentation in detail.

Nirusu replied:
Documentation can be found in the marblerun.sh branch: https://github.com/edgelesssys/marblerun.sh/tree/feat/graphene-so

thomasten replied:
Makes sense. For the same reason we don't need it in the nginx or other (future) advanced samples.

Nirusu commented:
Applied most of the suggested changes, renamed both methods, added some ANSI color coding to highlight what part of the output is argv and which is envp... 

Think this is ready for a re-review and we can discuss the things I left unresolved :)

thomasten commented on samples/graphene-hello/spawn/hello.c:
```diff
@@ -0,0 +1,16 @@
+
+  puts("\n\033[0;33mEnvironment variables:\033[0m");
+
+  for (int i = 0; envp[i]; i++) {
```
Now that I tested it, I think we should only print the env var from the manifest (with getenv).

Nirusu replied:
I opted for a middle way and decided to use the old approach we used in EGo: Filter everything with "EDG_" and "MARBLE_PREDEFINED", so that any user defined variables (+ the ones defined in the Graphene manifest) are visible.

Honestly, I think showing every environment variable is not too bad in general. It's looks cluttered, yes, but gives the developer an insight into the existing environment of the Graphene.

However, you can check out my new approach for this. Guess that should be a compromise...? :)

thomasten replied:
I don't like the filtering code here, but agree that showing every env var may be good. How about printing only the first n chars?

thomasten commented on samples/graphene-hello/preload/manifest.json:
```diff
@@ -9,6 +9,9 @@
+                    "HELLO_WORLD": "Hello World from Marblerun!"
+                },
                 "Argv": [
                     "foo",
```
we should probably remove this

Nirusu replied:
Wouldn't it make sense to have env vars in both of the examples, not only just one?

thomasten replied:
Absolutely. I wanted to comment on line 16, but it's hard to see the difference.

Nirusu commented on CMakeLists.txt:
```diff
@@ -6,6 +6,9 @@ find_package(OpenEnclave CONFIG REQUIRED)
 if (NOT CMAKE_BUILD_TYPE)
   set(CMAKE_BUILD_TYPE Debug)
 endif ()
+if (NOT CMAKE_BUILD_TYPE STREQUAL Debug)
```
Guess it's better to split this into an own PR then?

Nirusu replied:
Anyway, added the commits here. Still can decide on how to merge this later on.

thomasten replied:
Fine with me. You should squash into the trimpath commit before merge.

thomasten commented on samples/graphene-hello/preload/hello.manifest.template:
```diff
@@ -0,0 +1,45 @@
+loader.preload = "file:$(GRAPHENEDIR)/Runtime/libsysdb.so"
+
+# run Marblerun premain
+loader.env.LD_PRELOAD = "./premain-graphene.so"
+loader.env.LD_LIBRARY_PATH = "/lib"
```
```suggestion
loader.preload = "file:$(GRAPHENEDIR)/Runtime/libsysdb.so"
loader.env.LD_LIBRARY_PATH = "/lib"

# run Marblerun premain
loader.env.LD_PRELOAD = "./premain-graphene.so"
```

thomasten commented on dockerfiles/Dockerfile.coordinator:
```diff
@@ -6,7 +6,7 @@ RUN git clone https://github.com/edgelesssys/marblerun.git /coordinator
 COPY --from=pull /coordinator /coordinator
 WORKDIR /coordinator/build
-RUN cmake ..
+RUN cmake -DCMAKE_BUILD_TYPE=RelWithDebInfo ..
```
we should also add this to the injector Dockerfile
